aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -98,18 +103,46 @@ compute_result_type(at::ArrayRef<OperandInfo> operands,
 
 std::tuple<Device, ScalarType> TensorIterator::compute_common_type() {
   // See [Result type computation] in TensorIterator.h
+
   auto result_type =
       compute_result_type(operands_,
-        [](const Tensor& t) { return t.dim() > 0; },
-        [](const Tensor& t) { return !t.unsafeGetTensorImpl()->is_wrapped_number(); },
-        [](const Tensor& t) { return true; });
+        [](const OperandInfo& op) { return op.tensor.dim() > 0; },
+        [](const OperandInfo& op) { return !op.tensor.unsafeGetTensorImpl()->is_wrapped_number(); },
+        [](const OperandInfo& op) { return true; });
+
+  // if non-zero-dim tensor result is an integral type and there's a zero-dim
+  // floating point operand, we'll promote the the floating point type.
+  if (isIntegralType(std::get<1>(result_type))) {
+    auto alternate = compute_result_type(operands_,
+        [](const OperandInfo& op) {
+          return isFloatingType(op.tensor.scalar_type()) && op.tensor.dim() == 0;
+        }
+    );
+    if (std::get<1>(alternate) != ScalarType::Undefined) {
+      // preserve device from original result
+      return std::make_tuple(std::get<0>(result_type), std::get<1>(alternate));
+    }
+  }
 
   TORCH_INTERNAL_ASSERT(std::get<1>(result_type) != ScalarType::Undefined);
   return result_type;
 }
 
+// just checks for safe promotion.
```

Comment:
I'd interpret this comment as being equivalent to `np.can_cast(...,casting='safe')`, but this isn't checking safe casting in that respect.

Some examples that differ:
```
>>> np.can_cast(np.float64, np.float32)
False

>>> np.can_cast(np.int64, np.int32)
False

>>> np.can_cast(np.uint8, np.int8)
False
```

Reply:
ah, yes, we still have some slight differences but also that comment is stale. In an earlier revision I had rejected double->float and long->int.